File: pyskynet/lualib/pyskynet/cluster.lua
local skynet = require("skynet")
local foreign_seri = require("pyskynet.foreign_seri")
local remotepack = foreign_seri.remotepack
local remoteunpack = foreign_seri.remoteunpack

local clusterd
local cluster = {}
local sender = {}
local task_queue = {}

local function request_sender(q, node)
	local ok, c = pcall(skynet.call, clusterd, "lua", "sender", node)
	if not ok then
		skynet.error(c)
		c = nil
	end
	-- run tasks in queue
	local confirm = coroutine.running()
	q.confirm = confirm
	q.sender = c
	for _, task in ipairs(q) do
		if type(task) == "table" then
			if c then
				skynet.send(c, "lua", "push", task[1], remotepack(table.unpack(task, 2, task.n)))
			end
		else
			skynet.wakeup(task)
			skynet.wait(confirm)
		end
	end
	task_queue[node] = nil
	sender[node] = c
end

local function get_queue(t, node)
	local q = {}
	t[node] = q
	skynet.fork(request_sender, q, node)
	return q
end

setmetatable(task_queue, { __index = get_queue } )

local function get_sender(node)
	local s = sender[node]
	if not s then
		local q = task_queue[node]
		local task = coroutine.running()
		table.insert(q, task)
		skynet.wait(task)
		skynet.wakeup(q.confirm)
		return q.sender
	end
	return s
end

function cluster.call(node, address, ...)
	-- foreign.remotepack(...) will free by cluster.core.packrequest
	return remoteunpack(skynet.rawcall(get_sender(node), "lua", skynet.pack("req",  address, remotepack(...))))
end

function cluster.send(node, address, ...)
	-- push is the same with req, but no response
	local s = sender[node]
	if not s then
		table.insert(task_queue[node], table.pack(address, ...))
	else
		skynet.send(sender[node], "lua", "push", address, remotepack(...))
	end
end

function cluster.open(port)
	assert(type(port) == "number", "cluster.open's args must be number")
	skynet.call(clusterd, "lua", "listen", "0.0.0.0", port)
end

function cluster.reload(config)
	skynet.call(clusterd, "lua", "reload", config)
end

function cluster.proxy(node, name)
	return skynet.call(clusterd, "lua", "proxy", node, name)
end

function cluster.register(name, addr)
	assert(type(name) == "string")
	assert(addr == nil or type(addr) == "number")
	return skynet.call(clusterd, "lua", "register", name, addr)
end

function cluster.query(node, name)
	return skynet.call(get_sender(node), "lua", "req", 0, skynet.pack(name))
end

skynet.init(function()
	clusterd = skynet.uniqueservice("foreign_clusterd")
end)

return cluster

File: pyskynet/cluster.py


import pyskynet

import pyskynet.boot
import pyskynet.proto as pyskynet_proto
from pyskynet.skynet_py_foreign_seri import remotepack, remoteunpack, luapack
import gevent
import gevent.event

clusterd = None


def __load_clusterd():
    global clusterd
    clusterd = pyskynet.uniqueservice("foreign_clusterd")


pyskynet.boot.init(__load_clusterd)

node_to_sender = {}
node_to_task_queue = {}


def __request_sender(task_queue, node):
    sender = pyskynet_proto.call(clusterd, "lua", "sender", node)[0]
    i = 0
    while i < len(task_queue):
        task = task_queue[i]
        # tuple for send task, AsyncResult for call task
        if type(task) == tuple:
            pyskynet_proto.send(sender, "lua", "push", task[0], remotepack(*task[1]))
        else:
            confirm = gevent.event.Event()
            task.set((confirm, sender))
            confirm.wait()
        i += 1
    node_to_task_queue.pop(node)
    node_to_sender[node] = sender


def __get_sender(node):
    sender = node_to_sender.get(node)
    if sender is None:
        task = gevent.event.AsyncResult()
        task_queue = node_to_task_queue.get(node)
        if task_queue is None:
            task_queue = [task]
            node_to_task_queue[node] = task_queue
            gevent.spawn(__request_sender, task_queue, node)
        else:
            task_queue.append(task)
        confirm, sender = task.get()
        confirm.set()
        return sender
    else:
        return sender


def call(node, address, *args):
    return remoteunpack(*pyskynet_proto.rawcall(
        __get_sender(node), "lua", *luapack("req", address, *remotepack(*args))))


def send(node, address, *args):
    sender = node_to_sender.get(node)
    if sender is None:
        node_to_task_queue[node].append((address, args))
    else:
        pyskynet_proto.send(sender, "lua", "push", address, remotepack(*args))


def open(port):
    assert 0 <= port <= 65535, "cluster.open's port must be in [0, 65535]"
    pyskynet_proto.call(clusterd, "lua", "listen", "0.0.0.0", port)
